types/qs/index.d.ts

```diff
@@ -55,6 +55,8 @@ declare namespace QueryString {
         interpretNumericEntities?: boolean;
     }
 
+    interface ParsedQs { [key: string]: string | string[] | ParsedQs }
+
     function stringify(obj: any, options?: IStringifyOptions): string;
-    function parse(str: string, options?: IParseOptions): any;
+    function parse(str: string, options?: IParseOptions): ParsedQs;
```

Comment:
`parse` can also return an array of any JS value; ie, `unknown` is a better choice here since TS doesn't have a valid "actual JS object" type.

Reply:
> `parse` can also return an array of any JS value

Interesting, TIL! Can you share an example of how that would happen? I'm curious if we can use a widen union type instead of `unknown`.

Reply:
Well, for one, "any JS value" includes "an object that can be indexed with symbols or bigints", something TS is simply incapable of representing.

In particular, you can pass a custom decoder, and that can return anything. `qs.parse('a=1', { decoder() { return arguments; } })` is a fun easy one to play with :-)

Any wide union type would have to include `unknown`, or else it would be incorrect.

Reply:
> In particular, you can pass a custom decoder, and that can return anything.

Ah, I see. Thanks for clarifying.

My thinking here was that the return type would model the default and most frequent behaviour, since that probably covers the majority of use cases where users do not customise the decoding behaviour (presumably). For all other users who are customising the decoding behaviour, they could use a type assertion:


```ts
const result = qs.parse(str) as MyCustomParsedQs
```

Alternatively, we could define `parse` so that it receives a generic, which if defined will be used as the return type:

```ts
function parse<P = ParsedQs>(str: string, options?: IParseOptions): P;

const result = qs.parse(str) // $ExpectType ParsedQs
const result2 = qs.parse<MyCustomParsedQs>(str) // $ExpectType MyCustomParsedQs
```

IIUC, this is quite a popular pattern. For example, this is how request parameters are modelled in the Express types.

Another alternative (and probably the safest): try and automatically infer the return type from provided `decoder` function:

```typescript
interface IParseOptions<P> {
    decoder?: (str: string, defaultDecoder: defaultDecoder, charset: string, type: 'key' | 'value') => P | string;
}

interface ParsedQs { [key: string]: string | string[] | ParsedQs }

function parse(str: string, options?: Omit<IParseOptions<string>, 'decoder'>): ParsedQs;
function parse<P>(str: string, options?: IParseOptions<P>): { [key: string]: P };
```

Usage:

```ts
// Default
{
    const obj = qs.parse('a=c');
    obj; // $ExpectType ParsedQs
    obj.foo; // $ExpectType string | ParsedQs | string[]
}
// Custom decoder
{
    const obj = qs.parse('a=c', {
        decoder: (str, defaultDecoder, charset, type) => {
            switch (type) {
                case 'key': return str;
                case 'value': return parseFloat(str);
            }
        },
    });
    obj; // $ExpectType { [key: string]: number; }
    obj.foo; // $ExpectType number
}
```

WDYT?

Reply:
Inferring the return type from the decoder function makes sense; however, unless the type of that function changes based on whether the `type` argument is `'key'` or `'value'`, you'd have to assume that values could be the same type as the key, or the value; for example, an object of string keys and symbol values, would imply a type of "an object of string or symbol keys, and string or symbol values".

Reply:
In that case, let's just use `{ [key: string]: unknown }` for the overload where a decoder is provided. I've updated this PR, please could you re-review?